File: src/main/java/org/minturtle/careersupport/user/service/UserService.java
package org.minturtle.careersupport.user.service;


import lombok.RequiredArgsConstructor;
import org.minturtle.careersupport.auth.utils.ApiTokenProvider;
import org.minturtle.careersupport.auth.utils.JwtTokenProvider;
import org.minturtle.careersupport.common.aop.Logging;
import org.minturtle.careersupport.common.exception.ConflictException;
import org.minturtle.careersupport.common.exception.UnAuthorizedException;
import org.minturtle.careersupport.user.dto.*;
import org.minturtle.careersupport.user.entity.User;
import org.minturtle.careersupport.user.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.Date;
import reactor.core.scheduler.Schedulers;

@Service
@RequiredArgsConstructor
@Logging
public class UserService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
    private final ApiTokenProvider apiTokenProvider;


    public Mono<UserInfoDto> registerUser(UserRegistrationRequest registrationDto) {
        return userRepository.findByUsername(registrationDto.getUsername())
                .flatMap(existingUser -> Mono.error(new ConflictException("Username already exists")))
                .switchIfEmpty(Mono.defer(() -> {
                    return Mono.fromCallable(() -> passwordEncoder.encode(registrationDto.getPassword()))
                            .subscribeOn(Schedulers.boundedElastic())
                            .flatMap(encodedPassword -> {
                                User user = User.builder()
                                        .nickname(registrationDto.getNickname())
                                        .username(registrationDto.getUsername())
                                        .password(encodedPassword)
                                        .build();
                                return userRepository.save(user);
                            });
                }))
                .map(UserInfoDto::of)
                .cast(UserInfoDto.class);
    }



    public Mono<UserLoginResponse> login(String username, String password) {
        return userRepository.findByUsername(username)
                .filter(user -> passwordEncoder.matches(password, user.getPassword()))
                .flatMap(user -> jwtTokenProvider.sign(UserInfoDto.of(user), new Date())
                        .map(token -> createUserLoginResponse(user, token)))
                .switchIfEmpty(Mono.error(new UnAuthorizedException("Invalid credentials")));
    }

    public Mono<UserInfoResponse> getUserInfo(String token){
        return jwtTokenProvider.verify(token)
                .map(UserInfoResponse::of);
    }

    private UserLoginResponse createUserLoginResponse(User user, String jwt){
        return new UserLoginResponse(user.getNickname(), jwt);
    }

    public Mono<UserApiAccessTokenResponse> getUserApiAccessToken(String id) {
        return userRepository.findById(id)
                .switchIfEmpty(Mono.error(new UnAuthorizedException("사용자를 조회할 수 없습니다.")))
                .flatMap(user -> apiTokenProvider.generate(UserInfoDto.of(user))
                        .flatMap(token -> {
                            user.setApiToken(token);
                            return userRepository.save(user).thenReturn(token);
                        })
                )
                .map(UserApiAccessTokenResponse::new);
    }
}

File: src/main/java/org/minturtle/careersupport/user/service/UserService.kt
package org.minturtle.careersupport.user.service

import kotlinx.coroutines.reactor.awaitSingle
import org.minturtle.careersupport.auth.utils.ApiTokenProvider
import org.minturtle.careersupport.auth.utils.JwtTokenProvider
import org.minturtle.careersupport.common.aop.Logging
import org.minturtle.careersupport.common.exception.ConflictException
import org.minturtle.careersupport.common.exception.UnAuthorizedException
import org.minturtle.careersupport.user.dto.*
import org.minturtle.careersupport.user.entity.User
import org.minturtle.careersupport.user.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.*


@Service
@Logging
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenProvider: JwtTokenProvider,
    private val apiTokenProvider: ApiTokenProvider
){

    suspend fun registerUser(registrationDto: UserRegistrationRequest): UserInfoDto {
        userRepository.findByUsername(registrationDto.username)?.let {
            throw ConflictException("Username already exists")
        }

        val encodedPassword = passwordEncoder.encode(registrationDto.password)

        val newUser = User.builder()
            .nickname(registrationDto.nickname)
            .username(registrationDto.username)
            .password(encodedPassword)
            .build()

        val savedUser = userRepository.save(newUser).awaitSingle()

        return UserInfoDto.of(savedUser)
    }

    suspend fun login(username: String, password: String): UserLoginResponse {
        val user = userRepository.findByUsername(username)
            ?.takeIf { passwordEncoder.matches(password, it.password) }
            ?: throw UnAuthorizedException("Invalid credentials")

        val jwt = jwtTokenProvider.sign(UserInfoDto.of(user), Date()).awaitSingle()
        return UserLoginResponse.of(user, jwt)
    }

    suspend fun getUserInfo(token: String): UserInfoResponse {
        val verifiedUserInfo = jwtTokenProvider.verify(token).awaitSingle()

        return UserInfoResponse.of(verifiedUserInfo)
    }

    suspend fun getUserApiAccessToken(id: String): UserApiAccessTokenResponse {
        val findUser = userRepository.findById(id).awaitSingle()

        val token = apiTokenProvider.generate(UserInfoDto.of(findUser)).awaitSingle()

        findUser.apiToken = token

        userRepository.save(findUser).awaitSingle()
        return UserApiAccessTokenResponse(token)
    }
}